### PR TITLE
Make read_version_from_raw() more robust

### LINKED ISSUE #27: ValueError: embedded null byte
I get `ValueError: embedded null byte` when running `eclean-kernel` version 2.99.3:

```
* Removing kernel 5.13.16-gentoo-dist (not referenced by bootloader (grub2))                                                                                                             
Traceback (most recent call last):                                                                        
  File "/usr/lib/python-exec/python3.9/eclean-kernel", line 33, in <module>                                                                                                              
    sys.exit(load_entry_point('eclean-kernel==2.99.3', 'console_scripts', 'eclean-kernel')())   
  File "/usr/lib/python3.9/site-packages/ecleankernel/__main__.py", line 382, in setuptools_main                                                                                         
    sys.exit(main(sys.argv[1:]))                                                                                                                                                         
  File "/usr/lib/python3.9/site-packages/ecleankernel/__main__.py", line 332, in main                                                                                                    
    p = subprocess.Popen(scmd)                                                                            
  File "/usr/lib/python3.9/subprocess.py", line 951, in __init__                                 self._execute_child(args, executable, preexec_fn, close_fds,                                                                                                                         
  File "/usr/lib/python3.9/subprocess.py", line 1754, in _execute_child                                                                                                                  
    self.pid = _posixsubprocess.fork_exec(     
ValueError: embedded null byte
```

Using `pdb` shows the null byte:

```
* Removing kernel 5.13.15-gentoo-dist (not referenced by bootloader (grub2))
> /usr/lib/python3.9/site-packages/ecleankernel/__main__.py(332)main()
-> p = subprocess.Popen(scmd)
(Pdb) 
 [-] /boot/vmlinuz-5.13.15-gentoo-dist
 [-] /lib/modules/5.13.15-gentoo-dist
 [-] /boot/System.map-5.13.15-gentoo-dist
 [-] /boot/config-5.13.15-gentoo-dist
* Removing kernel 5.13.15-gentoo-dist (not referenced by bootloader (grub2))
> /usr/lib/python3.9/site-packages/ecleankernel/__main__.py(332)main()-> p = subprocess.Popen(scmd)
(Pdb) p scmd
['kernel-install', 'remove', '\x00CIFS', '/boot/initramfs-5.13.15-gentoo-dist.img']
(Pdb) p kf.internal_version
'\x00CIFS'
(Pdb) p kf
KernelImage(PosixPath('/boot/initramfs-5.13.15-gentoo-dist.img'))
```

Maybe a bug in `KernelImage.read_internal_version()`?

### PR DESCRIPTION
**Pull request by Flowdalic**
Ignore the "parsed" Linux kernel version information if there are non-ASCII at the beginning of the selected buffer.

Fixes #27.

### CODE REVIEW

**Comment by mgorny** [comment]
I suppose this would work most of the time, though I would prefer to know why some files don't have the version there. I suspect you may still accidentally grab some random string as version.

**Comment by Flowdalic** [comment]
> though I would prefer to know why some files don't have the version there.

Likely because they are not a Linux kernel. As I wrote in https://github.com/projg2/eclean-kernel/issues/27#issuecomment-1759625817 that happens when eclean-kernel looks at an **initramfs**.

For example, `gunzip initramfs-6.1.38-gentoo-dist.img -c | strings |less` shows

```
3CIFS: VFS: cifs ioctl 0x%x
7CIFS: cifs ioctl 0x%x
7CIFS: ioctl notify rc %d
7CIFS: unsupported ioctl
cifs
Linux version 
CIFS VFS Client for Linux
fs/smb/client/sess.c
OS/2
CIFS: %s: OS/2 server
```

**Comment by Jannik2099** [comment]
hmm, do the raw, decompressed kernel images share some magic? Are they guaranteed to be ELF files on all arches?

**Comment by mgorny** [comment]
I hoped you'd have some idea.

**Comment by mgorny** [comment]
CC @AndrewAmmerlaan too.

**Comment by AndrewAmmerlaan** [comment]
> Are they guaranteed to be ELF files on all arches?

No, the `vmlinuz.efi` on arm64 and riscv (with `CONFIG_EFI_ZBOOT`) are PE32+ not ELF. Maybe some other formats are also possible with different config options.

**Comment by AndrewAmmerlaan** [comment]
> CC @AndrewAmmerlaan too.

I'm not sure I understand the original issue this is fixing, what non-ASCII characters are there and how did they get there?

**Comment by Flowdalic** [comment]
> I'm not sure I understand the original issue this is fixing, what non-ASCII characters are there and how did they get there?

tl;dr: The match logic employed by  `read_version_from_raw()` is fragile, as it assumes that there is *always* a version String right after the String `Linux version `. This fails if the binary just has the null-byte terminated string `Linux version ` embedded. Which is not unlikely.

The `read_version_from_raw()` function performs a naive search for the String `Linux version ` on a given binary. If it matches, then it assumes that what follows the match is another string containing the version number. However, any binary may contain the String `Linux version \0`, after which arbitrary bytes follow. This is the case for the `initramfs` binary that triggers this. Here, `read_version_from_raw()` returns a byte array starting with the null byte and some more bytes. This byte array is later passed to `str()`, which also seems to work fine. However, once the resulting string is passed to `subprocess.Popen()`, because it wants to invoke `'kernel-install', 'remove', <kernel-version>, <kernel-filename>`, it throws, because <kernel-version> is a String that starts with the null byte: `'\x00CIFS'`.

**Comment by mgorny** [comment]
@Flowdalic, btw doesn't it incorrectly recognize your initramfs as kernel then?

**Comment by Flowdalic** [comment]
> @Flowdalic, btw doesn't it incorrectly recognize your initramfs as kernel then?

That is correct.

**Comment by mgorny** [comment]
@AndrewAmmerlaan, the problem is that the original logic combined getting internal version with checking file magic for kernel. However, as we added support for "raw" kernels, we've ended up looking for strings without any extra magic verification, so we end up treating any file containing `Linux version` as a kernel.

My idea is that we ought to require some kind of magic for "raw" kernels.

**Comment by Flowdalic** [comment]
> My idea is that we ought to require some kind of magic for "raw" kernels.

That would be ideal, if feasible. In addition (or in the meantime), can we get this PR merged to make things more robust?

**Comment by ajakk** [comment]
Then how about tacking on something like this, to make files with invalid magic be percieved as invalid kernels in the caller?
```
diff --git a/ecleankernel/file.py b/ecleankernel/file.py
index 1a79c9f..39506b4 100644
--- a/ecleankernel/file.py
+++ b/ecleankernel/file.py
@@ -205,6 +205,10 @@ class KernelImage(GenericFile):

         # check if it's compressed first
         b = self.decompress_raw(f)
+
+        if not b.startswith(b"MZ"):
+            return None
+
         # unlike with bzImage, the raw kernel binary has no header
         # that includes the version, so we parse the version message
         # that appears on boot
```

Seems to work for me at least, but I've not tested it everywhere available to me. Will need some tests to have magic added to their test kernel files, too, I think.

**Comment by Flowdalic** [comment]
I am still running into this.

Could we please get a fix into eclean-kernel?

If ajakk's suggestion of using `MZ` as magic byte is robust, then we can go obviously with that. However, if nobody knows if it is robust, then please consider merging this PR, even if it is not the perfect solution. I am fairly certain that it is pretty robust, resolves the issue and it certainly has no false negatives.

**Comment by storm37000** [comment]
tried your changed file.py and it totally breaks the entire program with `SystemError('No vmlinuz found. This seems ridiculous, aborting.')`

**Comment by Flowdalic** [comment]
> tried your changed file.py and it totally breaks the entire program with `SystemError('No vmlinuz found. This seems ridiculous, aborting.')`

I don't immediately see how this could be caused by the change proposed in this PR. Of course, I can not rule it out. On the other hand, I have been using the PR's patch for a while without issues.

Could you maybe provide more information and fire up pdb and see why it breaks?
